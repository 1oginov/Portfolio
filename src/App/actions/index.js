/* @flow */

import * as T from './types';

export const navigate = (route: string, params?: {}) => dispatch => {
  dispatch({
    payload: { route, params },
    type: T.NAVIGATE,
  });
};
