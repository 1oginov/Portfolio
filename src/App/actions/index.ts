import { Action } from 'redux';
import { ThunkAction } from 'redux-thunk';

import State from 'State';

import { navigationHappened } from './creators';
import { RouteParamsState } from '../State';
import * as R from '../routes';

export interface NavigateAction {
  (route: R.Route, params?: RouteParamsState): void;
}

export const navigate = (
  route: R.Route,
  params?: RouteParamsState,
): ThunkAction<void, State, null, Action> => dispatch => {
  dispatch(navigationHappened(route, params));
};
