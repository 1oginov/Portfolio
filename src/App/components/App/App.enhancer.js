/* @flow */

import withStockQuotesUpdater from 'Stocks/enhancers/withStockQuotesUpdater';

export default withStockQuotesUpdater;
