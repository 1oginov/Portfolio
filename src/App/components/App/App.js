/* @flow */

import * as React from 'react';

import Portfolio from 'Portfolio/components/Porfolio';
import View from 'ZenMoney/components/View';

const App = () => (
  <React.Fragment>
    <View />
    <Portfolio />
  </React.Fragment>
);

export default App;
