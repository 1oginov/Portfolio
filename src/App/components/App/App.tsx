import * as React from 'react';

import Navigation from '../Navigation';
import Guest from '../../views/Guest';
import Loading from '../../views/Loading';

interface Props {
  isAuthenticated: boolean;
  progress: boolean;
}

const App = ({ isAuthenticated, progress }: Props) => {
  if (progress) {
    return <Loading />;
  }

  if (!isAuthenticated) {
    return <Guest />;
  }

  return <Navigation />;
};

export default App;
