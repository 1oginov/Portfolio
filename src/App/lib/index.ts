import { annualizePercent, Position } from 'Portfolio/lib';
import { getQuotePrice } from 'Stocks/lib';
import Quote from 'Stocks/lib/IEX/Quote';

// TODO: Tests.

export interface StockPosition extends Position {
  openSum: number;

  closeSum: number | null;
  closePL: number | null;
  closePLPercent: number | null;
  closePLAnnualPercent: number | null;

  quote: Quote | null;
  quoteProgress: boolean;

  companyName: string | null;
  dailyPL: number | null;
  dailyPLPercent: number | null;

  marketPrice: number | null;
  marketSum: number | null;
  marketPL: number | null;
  marketPLPercent: number | null;
  marketPLAnnualPercent: number | null;
}

export interface OpenPositionsSummary {
  symbol: string;
  amount: number;

  openPriceAverage: number;
  openSum: number;

  companyName: string | null;
  dailyPL: number | null;
  dailyPLPercent: number | null;

  marketPL: number | null;
  marketPLPercent: number | null;
}

export interface Totals {
  totalOpenSum: number;

  totalCloseSum: number;
  totalClosePL: number;
  totalClosePLPercent: number;

  totalDailyPL: number;
  totalDailyPLPercent: number;

  totalMarketPL: number;
  totalMarketPLPercent: number;
  totalMarketSum: number;
}

export const createStockPosition = (position: Position, quote: Quote | null, quoteProgress: boolean): StockPosition => {
  const openSum = position.amount * position.openPrice + position.openCommission;

  let closeSum = null;
  let closePL = null;
  let closePLPercent = null;
  let closePLAnnualPercent = null;

  if (position.closeCommission !== null && position.closeDate !== null && position.closePrice !== null) {
    closeSum = position.amount * position.closePrice - position.closeCommission;
    closePL = closeSum - openSum;
    closePLPercent = closePL / openSum;
    closePLAnnualPercent = annualizePercent(closePLPercent, position.openDate, position.closeDate);
  }

  let companyName = null;
  let dailyPL = null;
  let dailyPLPercent = null;

  if (quote !== null) {
    ({ companyName } = quote);
    dailyPL = position.amount * quote.change;
    dailyPLPercent = quote.changePercent;
  }

  const marketPrice = getQuotePrice(quote);
  let marketSum = null;
  let marketPL = null;
  let marketPLPercent = null;
  let marketPLAnnualPercent = null;

  if (marketPrice !== null) {
    marketSum = position.amount * marketPrice;
    marketPL = marketSum - openSum;
    marketPLPercent = marketPL / openSum;
    marketPLAnnualPercent = annualizePercent(marketPLPercent, position.openDate);
  }

  return {
    ...position,
    closePL,
    closePLAnnualPercent,
    closePLPercent,
    closeSum,
    companyName,
    dailyPL,
    dailyPLPercent,
    marketPL,
    marketPLAnnualPercent,
    marketPLPercent,
    marketPrice,
    marketSum,
    openSum,
    quote,
    quoteProgress,
  };
};

export const createOpenPositionsSummaries = (stockPositions: StockPosition[]): OpenPositionsSummary[] => {
  const bySymbols: { [key: string]: StockPosition[] } = {};

  stockPositions.forEach(position => {
    if (!bySymbols[position.symbol]) {
      bySymbols[position.symbol] = [];
    }

    bySymbols[position.symbol].push(position);
  });

  return Object.keys(bySymbols).map(symbol => {
    let amount = 0;
    let openPrice = 0;
    let openSum = 0;
    let dailyPL: number | null = null;
    let dailyPLPercent: number | null = null;
    let marketPL: number | null = null;
    let marketPLPercent: number | null = null;

    bySymbols[symbol].forEach(stockPosition => {
      amount += stockPosition.amount;
      openPrice += stockPosition.amount * stockPosition.openPrice;
      openSum += stockPosition.openSum;

      if (stockPosition.dailyPL !== null) {
        if (dailyPL === null) {
          dailyPL = 0;
        }

        dailyPL += stockPosition.dailyPL;
      }

      if (stockPosition.marketPL !== null) {
        if (marketPL === null) {
          marketPL = 0;
        }

        marketPL += stockPosition.marketPL;
      }
    });

    const referencePosition = bySymbols[symbol][0];

    if (dailyPL !== null && referencePosition.quote !== null) {
      dailyPLPercent = dailyPL / amount / referencePosition.quote.close;
    }

    if (marketPL !== null) {
      marketPLPercent = marketPL / openSum;
    }

    const openPriceAverage = openPrice / amount;
    const { companyName } = referencePosition;

    return {
      amount,
      companyName,
      dailyPL,
      dailyPLPercent,
      marketPL,
      marketPLPercent,
      openPriceAverage,
      openSum,
      symbol,
    };
  });
};

export const calculateTotals = (stockPositions: StockPosition[]): Totals => {
  let totalOpenSum = 0;
  let totalCloseSum = 0;
  let totalDailyPL = 0;
  let totalPreviousCloseSum = 0;
  let totalMarketSum = 0;

  stockPositions.forEach(position => {
    totalOpenSum += position.openSum;

    if (position.closeSum !== null) {
      totalCloseSum += position.closeSum;
    }

    if (position.quote !== null) {
      totalDailyPL += position.amount * position.quote.change;
      totalPreviousCloseSum += position.amount * position.quote.close;
    }

    if (position.marketSum !== null) {
      totalMarketSum += position.marketSum;
    }
  });

  const totalClosePL = totalCloseSum - totalOpenSum;
  const totalClosePLPercent = totalClosePL / totalOpenSum;

  const totalDailyPLPercent = totalDailyPL / totalPreviousCloseSum;

  const totalMarketPL = totalMarketSum - totalOpenSum;
  const totalMarketPLPercent = totalMarketPL / totalOpenSum;

  return {
    totalClosePL,
    totalClosePLPercent,
    totalCloseSum,
    totalDailyPL,
    totalDailyPLPercent,
    totalMarketPL,
    totalMarketPLPercent,
    totalMarketSum,
    totalOpenSum,
  };
};
