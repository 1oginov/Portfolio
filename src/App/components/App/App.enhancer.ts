import { connect } from 'react-redux';
import { isEmpty, isLoaded } from 'react-redux-firebase';
import { branch, compose, withProps } from 'recompose';

import { AuthState } from 'Firebase/State';
import withStockQuotesUpdater from 'Stocks/enhancers/withStockQuotesUpdater';
import State from 'State';

import { Props } from './App';

interface StateProps {
  auth: AuthState;
}

const mapStateToProps = ({ firebase: { firebase: { auth } } }: State): StateProps => ({ auth });

export default compose<Props & StateProps, {}>(
  connect(mapStateToProps),
  withProps<Props, StateProps>(({ auth }) => ({
    isAuthenticated: !isEmpty(auth) && isLoaded(auth),
    progress: !isLoaded(auth),
  })),
  branch<Props & StateProps>(
    ({ isAuthenticated }) => isAuthenticated,
    withStockQuotesUpdater,
  ),
);
