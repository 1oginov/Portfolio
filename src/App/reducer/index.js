/* @flow */

import * as T from '../actions/types';
import * as R from '../routes';

const initialState = {
  route: R.HOME,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case T.NAVIGATE:
      return {
        ...state,
        route: action.payload,
      };

    default:
      return state;
  }
};
