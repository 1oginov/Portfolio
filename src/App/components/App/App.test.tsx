import { shallow } from 'enzyme';
import * as React from 'react';

import App from './App';

it('matches snapshot if loading', () => {
  const wrapper = shallow(<App isAuthenticated={false} progress />);

  expect(wrapper).toMatchSnapshot();
});

it('matches snapshot if unauthenticated', () => {
  const wrapper = shallow(<App isAuthenticated={false} progress={false} />);

  expect(wrapper).toMatchSnapshot();
});

it('matches snapshot if authenticated', () => {
  const wrapper = shallow(<App isAuthenticated progress={false} />);

  expect(wrapper).toMatchSnapshot();
});
