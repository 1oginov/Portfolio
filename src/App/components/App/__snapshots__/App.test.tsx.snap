// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot if authenticated 1`] = `<Connect(Navigation) />`;

exports[`matches snapshot if loading 1`] = `<Loading />`;

exports[`matches snapshot if unauthenticated 1`] = `<Guest />`;
